### ISSUE TITLE
Can `Var` have an `id` field?

### ISSUE THREAD

**Issue by cscherrer** [JuliaStaging/JuliaVariables.jl#28]
For code rewriting, it can be really useful to have more to distinguish a variable than just its name. For example, we might want to do things like "rename all occurrences of this `x` to `y`" or even "convert this AST to SSA style". 

This could be very easy if JuliaVariables could be extended so `Var` has an additional `id` field. Do you have how difficult that would be to add? If it would be a major rewrite I could try to find other ways of getting to this.

**Comment by thautwarm** [comment]
It's possible, as `SymRef` should be its object identity. We can assign id to Var at

https://github.com/JuliaStaging/JuliaVariables.jl/blob/67091c1506e5296b9f6ab7b62b7e58c93fa1e3a0/src/JuliaVariables.jl#L479

the same `SymRef` creates the same id in `Var`.

NOTE that the global variables and keyword arguments cannot be refactored by JuliaVariables.

**Comment by cscherrer** [comment]
Ok, I have at least a start on this. I'll PR soon to see what you think.

**Comment by cscherrer** [comment]
Hmm, maybe `unique_id` should be the same as `name` for globals, and we only increment the counter for local variables?

**Comment by thautwarm** [comment]
Sorry, `SymRef` is not the object identity of the resolved symbols. We should handle `LocalVar` and `GlobalVar`(`Symbol`). 

I think your modification to `local_var_to_var` might not work for some cases. Same symbols in different scopes might result in different `id`.

Fortunately, `LocalVar.is_mutable` or `LocalVar.is_shared` are references, they can be the real object identity of the symbol. `NameResolution.jl` recognizes the same `LocalVar` for their different occurrences.

https://github.com/JuliaStaging/NameResolution.jl/blob/463ad09584ce823514a58bbc1391c6db51035a0b/src/Variable.jl#L2

You can use either `is_mutable` and `is_shared` as the reference. I'd suggest such a structure for `Var`:

```julia
struct Var
    name::Symbol
    unique_id:: Union{Ref{Bool}, Nothing}  # nothing when it's a global variable
    is_mutable::Bool
    is_shared::Bool
    is_global::Bool
end
```

**Comment by cscherrer** [comment]
Ok, I think I understand, but just to be sure... For two `LocalVar`s, are the `LocalVar.is_mutable`s the same (`===`) exactly when the  `LocalVar.is_shared`s are the same?

**Comment by thautwarm** [comment]
@cscherrer  Yes.. Sorry for such redundancy, Maybe I should make `LocalVar` a mutable struct then keeping a LocalVar is sufficent..

**Comment by cscherrer** [comment]
That would make sense, but it's not such a big deal either way. I think I'd refactor it this way if you think you'll be extending it more, since that could be easier to build on. But if it's basically done I'd maybe not worry about it. Up to you of course :)

**Comment by cscherrer** [comment]
Ok, I've updated the PR.

Also, what do you think of allowing another method passing an `id_function`? I've prototyped the idea here:
https://github.com/cscherrer/JuliaVariables.jl/pull/1
